src/utils.cpp

```diff
@@ -334,6 +335,621 @@ void cnf_to_dimacs(Term cnf, std::ostringstream & y)
+          vec.push_back(term_name);
+        }
+        Term term_name;
+        query_cache(term, term_name);
+
+        term_name = give_symbolic_name(term);  // making a new symbol
+        save_in_cache(term, term_name);
+
+        reduced.push_back({ term_name, solver_->make_term(op, vec) });
+      }
+      else
+      {  // mapping a symbolic constant to itself
+        save_in_cache(term, term);
+      }
+    }
+
+    return Walker_Continue;
+  }
+};
+
+// binarising xor with multiple children
+class XorBinarize : public IdentityWalker
+{
+ public:
+  XorBinarize(SmtSolver solver_) : IdentityWalker{ solver_, false } {}
+  WalkerStepResult visit_term(Term & term)
+  {
+    if (!preorder_)
+    {
+      smt::Op op = term->get_op();
+      if (!op.is_null())
+      {
+        if (op == smt::Xor)
+        {
+          auto it = term->begin();
+          Term term_name;
+          query_cache((*it),
+                      term_name);  // finding the mapped term from the cache
+          Term ne = term_name;
+          it++;
+          while (it != term->end())
+          {  // Binarising the term
+            Term term_name;
+            query_cache((*it), term_name);
+            ne = solver_->make_term(op, ne, term_name);
+            it++;
+          }
+          save_in_cache(term, ne);  // storing the new binarised term in cache
+        }
+        else
+        {
+          save_in_cache(term, term);
+        }
+      }
+      else
+      {
+        save_in_cache(term, term);
+      }
+    }
+
+    return Walker_Continue;
+  }
+  Term acc_cache(Term term)
+  {
+    Term ne;
+    query_cache(term, ne);
+    return ne;
+  }
+};
+
+// Given a boolean formula, removes terms "true" and "false" to give a formula
+// without adding new symbols
+class EliminateBooleanConstants : public IdentityWalker
+{
+ public:
+  EliminateBooleanConstants(SmtSolver solver_)
+      : IdentityWalker{ solver_, false }
+  {
+  }
+  WalkerStepResult visit_term(Term & term)
+  {
+    auto is_true = [&](Term t) {  // If the term is "true"
+      return ((t->to_string()) == "true");
+    };
+    auto is_false = [&](Term t) {  // If the term is "false"
+      return ((t->to_string()) == "false");
+    };
+    if (!preorder_)
+    {
+      smt::Op op = term->get_op();
+      if (!op.is_null())
+      {
+        Term tr = solver_->make_term(true);   // Term "true"
+        Term fa = solver_->make_term(false);  // Term "false"
+        if (op == smt::Not)
+        {
+          Term t = (*term->begin());
+          Term term_name;
+          query_cache(t,
+                      term_name);  // Querying the mapped formula of the child
+                                   // as we are doing a preorder traversal
+
+          if (is_true(term_name))
+          {  // not(false)=true
+            save_in_cache(term, fa);
+          }
+          else if (is_false(term_name))
+          {  // not(true)=false
+            save_in_cache(term, tr);
+          }
```

Comment:
It is unclear how `true` and `false` are eliminated here -- they are stored in the cache. Can you clarify inside a comment?